lib/logstash/config/config_ast.rb

```diff
@@ -231,7 +233,14 @@ def compile
       when "filter"
         return <<-CODE
           #{variable_name}.filter(event) {|new_event| events << new_event }
+
+          # Break early if the event was cancelled
+          return [] if events.all?(&:cancelled?)
```

Comment:
If this is to be merged with #3110,

line 235 will use the multi_filter method, so events will be empty if all events are cancelled, this could be changed to return events if events.empty?